Initial commit

Taras-Ivashchuk opened the pull request.

Taras-Ivashchuk committed:
Initial commit

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,31 @@
+        Person.people[name] = self
 
 
 def create_person_list(people: list) -> list:
```
The `Person.people` class attribute will retain its state across multiple calls to `create_person_list`. This can lead to unexpected results if you run the function with different sets of people. It's good practice to clear this dictionary at the beginning of the function to ensure each call is independent.

Taras-Ivashchuk replied:
Fixed

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,31 @@
-    # write your code here
-    pass
+    person_list = []
+    for person in people:
```
This violates checklist item #3: 'When creating a list of instances from a collection (such as a list of dictionaries), it is recommended to use **list comprehension** rather than traditional `for` loops.' A cleaner approach would be to first create all `Person` instances using a list comprehension, and then use a separate loop to set up the spouse relationships. This separates the concerns of object creation and object linking.

Taras-Ivashchuk replied:
Fixed

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,31 @@
+        if wife_name is not None:
+            # Search a wife for husband to make the pair
+            wife = Person.people.get(wife_name)
+            if wife:
```
This nested `if` statement violates checklist item #2: 'Avoid nested `if` by using `and`, `or` logical operators.' You can combine the check from line 16 with this one. For example, you could assign the result of `Person.people.get(wife_name)` to a variable and then check if that variable is truthy in a single `if` statement.

Taras-Ivashchuk replied:
Fixed

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Great work on implementing the `Person` class and the core logic for linking spouses. However, a few changes are needed before your solution can be approved.

The most critical issue is that the `Person.people` class attribute retains its state across multiple calls to `create_person_list`. This is a functional bug, as data from a previous call will interfere with the next, leading to incorrect behavior. To fix this, please clear the `Person.people` dictionary at the beginning of the function.

Additionally, your code deviates from the project checklist in two ways:
- You've used a nested `if` statement, which should be simplified into a single conditional using logical operators [CHECKLIST ITEM #2].
- The list of `Person` instances should be created using a list comprehension rather than a `for` loop for better readability and conciseness [CHECKLIST ITEM #3].

Once these issues are addressed, your code will be much more robust and aligned with the project's best practices.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

Taras-Ivashchuk committed:
1. Person.people cleared in the create_person_list.
2. Used list comprehension to create person_list
3. Improved conditional checks.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,24 @@
+    person_list = [Person(person["name"], person["age"]) for person in people]
+
+    for person in people:
+        if "wife" in person and person["wife"] is not None:
```
This violates checklist item #4: 'When accessing values in a dictionary, it is better to use the `dict.get()` method instead of explicitly checking for the presence of a key.' You can simplify this condition to `if person.get('wife'):`.

petrykivd replied:
Agree with AI, please read checklist berfore pushing your solution 😉

Taras-Ivashchuk replied:
Fixed

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,24 @@
+
+    for person in people:
+        if "wife" in person and person["wife"] is not None:
+            # Search a wife for husband to make the pair
```
According to the checklist, comments are not needed as the code should be self-explanatory. Please remove this comment.

Taras-Ivashchuk replied:
Comments removed

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,24 @@
+            # Search a wife for husband to make the pair
+            wife = Person.people[person["wife"]]
+            Person.people[person["name"]].wife = wife
+        elif "husband" in person and person["husband"] is not None:
```
This also violates checklist item #4. Similar to the check for 'wife', this condition can be simplified by using `person.get('husband')`.

petrykivd replied:
+

Taras-Ivashchuk replied:
Fixed

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,8 +1,24 @@
+            wife = Person.people[person["wife"]]
+            Person.people[person["name"]].wife = wife
+        elif "husband" in person and person["husband"] is not None:
+            # Search a husband for wife to make the pair
```
According to the checklist, comments are not needed as the code should be self-explanatory. Please remove this comment.

petrykivd replied:
+

Taras-Ivashchuk replied:
Comments removed

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Great progress on this! Your code's core logic is working perfectly, and you've addressed the major issues from the last review. I'm requesting a couple of small changes to fully align your code with the project's checklist.

- The project requires using the `dict.get()` method for dictionary access [CHECKLIST ITEM #4]. Instead of checking `if 'wife' in person:`, you can simplify this to `if person.get('wife'):`. Please apply this to both the 'wife' and 'husband' checks.
- The checklist also asks for self-explanatory code without comments. Please remove the comments inside the `for` loop.

These are quick fixes. Once you've made these adjustments, your solution will be ready for approval. Keep up the great work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

petrykivd commented on app/main.py:
```diff
@@ -1,8 +1,24 @@
 def create_person_list(people: list) -> list:
-    # write your code here
-    pass
+    Person.people = {}
```
Is it really necessary to clear the dictionary with people every time?

Taras-Ivashchuk replied:
@petrykivd hello! Sorry for delay. Each time I call create_person_list function with unique list creates the people dictionary that has  its own state that survive calls. That makes people dictionary too big, thus slowing down the look up.

petrykivd replied:
Searching the dictionary is fast, regardless of size, because it is a hash map (you will learn more about this later).

In general, there is no clear requirement in the task to clear this dictionary, so let it be as it is.

Taras-Ivashchuk replied:
Got it, fixed

Taras-Ivashchuk committed:
Removed comments
Instead of explicitly checking keys used dict.get() to access keys

petrykivd reviewed (approved):
Nice! 🌟

Taras-Ivashchuk committed:
Removed setting Person.people = {} as unnecessary
